In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
Final_Dataset = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Book Recommendation System/Final_Dataset.csv')
Final_Dataset.head()

Unnamed: 0  User-ID   Age Country        ISBN  Book-Rating  Avg_Rating  \
0           0        8  33.0  canada  0002005018            5    7.666667   
1           1    11676  28.0     NaN  0002005018            8    7.666667   
2           2    67544  30.0  canada  0002005018            8    7.666667   
3           3   116866  32.0   other  0002005018            9    7.666667   
4           4   123629  33.0  canada  0002005018            9    7.666667   

   Total_No_Of_Users_Rated    Book-Title           Book-Author  \
0                        9  Clara Callan  Richard Bruce Wright   
1                        9  Clara Callan  Richard Bruce Wright   
2                        9  Clara Callan  Richard Bruce Wright   
3                        9  Clara Callan  Richard Bruce Wright   
4                        9  Clara Callan  Richard Bruce Wright   

   Year-Of-Publication              Publisher  
0               2001.0  HarperFlamingo Canada  
1               2001.0  HarperFlamingo Canada  
2               2001.0  HarperFlamingo Canada  
3               2001.0  HarperFlamingo Canada  
4               2001.0  HarperFlamingo Canada

# <font style="color: red"> **Popularity Based Filtering**

<font color='voilet'> *As the name suggests Popularity based recommendation system works with the trend. It basically uses the items which are in trend right now. For example, if any book which is usually bought by every new user then there are chances that it may suggest that book to the user who just signed up.<br>
Book weighted avg formula:<br>
Weighted Rating(WR)=[vR/(v+m)]+[mC/(v+m)]<br>
where,<br>
v is the number of votes for the books;<br>
m is the minimum votes required to be listed in the chart;<br>
R is the average rating of the book; and<br>
C is the mean vote across the whole report.<br>
Now we find the values of v,m,R,C.*

In [ ]:
C= Final_Dataset['Avg_Rating'].mean()
m= Final_Dataset['Total_No_Of_Users_Rated'].quantile(0.90)
Top_Books = Final_Dataset.loc[Final_Dataset['Total_No_Of_Users_Rated'] >= m]
print(f'C={C} , m={m}')
Top_Books.shape

C=7.626700569505161 , m=64.0


(38570, 12)

<font color='voilet'> *Here we used 90th percentile as our cutoff. In other words, for a book to feature in the charts, it must have more votes than at least 90% of the books in the list.<br>
We see that there are 38570 books which qualify to be in this list. Now, we need to calculate our metric for each qualified book. To do this, we will define a function, weighted_rating() and define a new feature score, of which we’ll calculate the value by applying this function to our DataFrame of qualified books:*


In [ ]:
def weighted_rating(x, m=m, C=C):
    v = x['Total_No_Of_Users_Rated']
    R = x['Avg_Rating']
    return (v/(v+m) * R) + (m/(m+v) * C)


Top_Books['Score'] = Top_Books.apply(weighted_rating,axis=1)


#Sorting books based on score calculated above
Top_Books = Top_Books.sort_values('Score', ascending=False)

In [ ]:
#Keeping only one entry of each book
Top_Books=Top_Books.sort_values('Score', ascending=False).drop_duplicates('ISBN').sort_index()
cm=sns.light_palette('yellow',as_cmap=True)
#Sorting books based on score calculated above
Top_Books = Top_Books.sort_values('Score', ascending=False)

#Printing the top 20 books
Top_Books[['Book-Title', 'Total_No_Of_Users_Rated', 'Avg_Rating', 'Score']].reset_index(drop=True).head(20).style.background_gradient(cmap=cm)

<font color='red'> **The Popularity based recommender provide a general chart of recommended books to all the users. They are not sensitive to the interests and tastes of a particular user.**

In [ ]:
Top_Books.head(1)

Unnamed: 0  User-ID   Age Country        ISBN  Book-Rating  Avg_Rating  \
5954        5954   171011  13.0     usa  0439139597           10    9.262774   

      Total_No_Of_Users_Rated                                    Book-Title  \
5954                      137  Harry Potter and the Goblet of Fire (Book 4)   

        Book-Author  Year-Of-Publication   Publisher     Score  
5954  J. K. Rowling               2000.0  Scholastic  8.741835

In [ ]:
Books = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Book Recommendation System/Books.csv')
Top_Books = pd.merge(Top_Books, Books, on='ISBN')
Top_Books = Top_Books[['ISBN','Book-Title_x','Book-Author_x','Year-Of-Publication_x','Publisher_x', 'Total_No_Of_Users_Rated', 'Avg_Rating','Image-URL-L']]
print(Top_Books.shape)
Top_Books.head()

(333, 8)


ISBN                                       Book-Title_x  \
0  0439139597       Harry Potter and the Goblet of Fire (Book 4)   
1  059035342X  Harry Potter and the Sorcerer's Stone (Harry P...   
2  043935806X  Harry Potter and the Order of the Phoenix (Boo...   
3  0446310786                              To Kill a Mockingbird   
4  0439136369  Harry Potter and the Prisoner of Azkaban (Book 3)   

   Book-Author_x  Year-Of-Publication_x                 Publisher_x  \
0  J. K. Rowling                 2000.0                  Scholastic   
1  J. K. Rowling                 1999.0      Arthur A. Levine Books   
2  J. K. Rowling                 2003.0                  Scholastic   
3     Harper Lee                 1988.0  Little Brown &amp; Company   
4  J. K. Rowling                 2001.0                  Scholastic   

   Total_No_Of_Users_Rated  Avg_Rating  \
0                      137    9.262774   
1                      313    8.939297   
2                      206    9.033981   
3                      214    8.943925   
4                      133    9.082707   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0439139597.0...  
1  http://images.amazon.com/images/P/059035342X.0...  
2  http://images.amazon.com/images/P/043935806X.0...  
3  http://images.amazon.com/images/P/0446310786.0...  
4  http://images.amazon.com/images/P/0439136369.0...

In [ ]:
Top_Books.to_csv('/content/drive/MyDrive/Colab Notebooks/Book Recommendation System/Top_Books.csv')

# ***Create User Book Matrix***

In [4]:
from sklearn.model_selection import train_test_split

In [5]:
first_half, second_half = train_test_split(Final_Dataset, test_size=0.95, shuffle = False)

In [6]:
del(second_half)
del(Final_Dataset)
first_half.shape

(19192, 12)

In [7]:
first_half[first_half['User-ID']==8]

Unnamed: 0  User-ID   Age Country        ISBN  Book-Rating  Avg_Rating  \
0            0        8  33.0  canada  0002005018            5    7.666667   
9            9        8  33.0  canada  074322678X            5    5.000000   
10          10        8  33.0  canada  0887841740            5    5.000000   
11          11        8  33.0  canada  1552041778            5    5.000000   
12          12        8  33.0  canada  1567407781            6    6.000000   
13          13        8  33.0  canada  1575663937            6    6.000000   
14          14        8  33.0  canada  1881320189            7    7.000000   

    Total_No_Of_Users_Rated  \
0                         9   
9                         1   
10                        1   
11                        1   
12                        1   
13                        1   
14                        2   

                                           Book-Title           Book-Author  \
0                                        Clara Callan  Richard Bruce Wright   
9             Where You'll Find Me: And Other Stories           Ann Beattie   
10                                 The Middle Stories           Sheila Heti   
11                                           Jane Doe          R. J. Kaiser   
12       The Witchfinder (Amos Walker Mystery Series)     Loren D. Estleman   
13  More Cunning Than Man: A Social History of Rat...    Robert Hendrickson   
14                      Goodbye to the Buttermilk Sky          Julia Oliver   

    Year-Of-Publication                    Publisher  
0                2001.0        HarperFlamingo Canada  
9                2002.0                     Scribner  
10               2004.0        House of Anansi Press  
11               1999.0                   Mira Books  
12               1998.0     Brilliance Audio - Trade  
13               1999.0  Kensington Publishing Corp.  
14               1994.0               River City Pub

In [9]:
first_half['User-ID'].to_csv('/content/drive/MyDrive/Colab Notebooks/Book Recommendation System/first_row.csv')

In [10]:
matrix = first_half.pivot_table(index='User-ID', columns='Book-Title', values='Book-Rating')
matrix.head()

Book-Title  1984  2061: Odyssey Three  \
User-ID                                 
8            NaN                  NaN   
9            NaN                  NaN   
12           NaN                  NaN   
14           NaN                  NaN   
16           NaN                  NaN   

Book-Title  28 Barbary Lane : A \Tales of the City\" Omnibus"  2nd Chance  \
User-ID                                                                     
8                                                         NaN         NaN   
9                                                         NaN         NaN   
12                                                        NaN         NaN   
14                                                        NaN         NaN   
16                                                        NaN         NaN   

Book-Title  A Beautiful Mind : A Biography of John Forbes Nash, Jr.  \
User-ID                                                               
8                                                         NaN         
9                                                         NaN         
12                                                        NaN         
14                                                        NaN         
16                                                        NaN         

Book-Title  A Child Called \It\": One Child's Courage to Survive"  \
User-ID                                                             
8                                                         NaN       
9                                                         NaN       
12                                                        NaN       
14                                                        NaN       
16                                                        NaN       

Book-Title  A Christmas Story  A Cuisine of Leftovers  \
User-ID                                                 
8                         NaN                     NaN   
9                         NaN                     NaN   
12                        NaN                     NaN   
14                        NaN                     NaN   
16                        NaN                     NaN   

Book-Title  A Dangerous Place: California's Unsettling Fate (Images of America)  \
User-ID                                                                           
8                                                         NaN                     
9                                                         NaN                     
12                                                        NaN                     
14                                                        NaN                     
16                                                        NaN                     

Book-Title  A Density of Souls  ...  Women on the Case  \
User-ID                         ...                      
8                          NaN  ...                NaN   
9                          NaN  ...                NaN   
12                         NaN  ...                NaN   
14                         NaN  ...                NaN   
16                         NaN  ...                NaN   

Book-Title  World of Pies : A Novel  Writing Down the Bones  \
User-ID                                                       
8                               NaN                     NaN   
9                               NaN                     NaN   
12                              NaN                     NaN   
14                              NaN                     NaN   
16                              NaN                     NaN   

Book-Title  Wuthering Heights  Wuthering Heights (Penguin Popular Classics)  \
User-ID                                                                       
8                         NaN                                           NaN   
9                         NaN                                           NaN   
12                        NaN              

In [9]:
# all the books that are rated by the user 226745
matrix.loc[56][matrix.loc[56]>0]

Book-Title
Haveli (Laurel Leaf Books)    9.0
LONESOME DOVE                 7.0
Name: 56, dtype: float64

# Cosine Similarity

In [11]:
from sklearn.metrics.pairwise import cosine_similarity

In [12]:
df = matrix

df_filled = df.fillna(0)
original_non_nan_counts = df.notna().sum(axis=1)

# Step 2: Calculate sum of all values per row
sum_of_all = df_filled.sum(axis=1)

# Step 3: Calculate normalization factor only where sum is not zero
normalization_factor = original_non_nan_counts.div(sum_of_all.where(sum_of_all != 0))

# Step 4: Normalize only rows that had non-zero sums initially
normalized_df = df_filled.sub(df_filled.multiply(normalization_factor, axis=0), fill_value=0)

# Step 5: Ensure completely zero rows stay zero
normalized_df[sum_of_all == 0] = 0

# free matrix datafram
# del(matrix)

# Print the resulting DataFrame
normalized_df

Book-Title  1984  2061: Odyssey Three  \
User-ID                                 
8            0.0                  0.0   
9            0.0                  0.0   
12           0.0                  0.0   
14           0.0                  0.0   
16           0.0                  0.0   
...          ...                  ...   
278681       0.0                  0.0   
278798       0.0                  0.0   
278832       0.0                  0.0   
278843       0.0                  0.0   
278851       0.0                  0.0   

Book-Title  28 Barbary Lane : A \Tales of the City\" Omnibus"  2nd Chance  \
User-ID                                                                     
8                                                         0.0         0.0   
9                                                         0.0         0.0   
12                                                        0.0         0.0   
14                                                        0.0         0.0   
16                                                        0.0         0.0   
...                                                       ...         ...   
278681                                                    0.0         0.0   
278798                                                    0.0         0.0   
278832                                                    0.0         0.0   
278843                                                    0.0         0.0   
278851                                                    0.0         0.0   

Book-Title  A Beautiful Mind : A Biography of John Forbes Nash, Jr.  \
User-ID                                                               
8                                                         0.0         
9                                                         0.0         
12                                                        0.0         
14                                                        0.0         
16                                                        0.0         
...                                                       ...         
278681                                                    0.0         
278798                                                    0.0         
278832                                                    0.0         
278843                                                    0.0         
278851                                                    0.0         

Book-Title  A Child Called \It\": One Child's Courage to Survive"  \
User-ID                                                             
8                                                         0.0       
9                                                         0.0       
12                                                        0.0       
14                                                        0.0       
16                                                        0.0       
...                                                       ...       
278681                                                    0.0       
278798                                                    0.0       
278832                                                    0.0       
278843                                                    0.0       
278851                                                    0.0       

Book-Title  A Christmas Story  A Cuisine of Leftovers  \
User-ID                                                 
8                         0.0                     0.0   
9                         0.0                     0.0   
12                        0.0                     0.0   
14                        0.0                     0.0   
16                        0.0                     0.0   
...                       ...                     ...   
278681                    0.0                     0.0   
278798                    0.0                     0.0   
278832                    0.0                     0.0   
278843         

In [13]:
normalized_df[normalized_df.index==8]['Goodbye to the Buttermilk Sky']

User-ID
8    5.74359
Name: Goodbye to the Buttermilk Sky, dtype: float64

In [14]:
cosine_distance = cosine_similarity(normalized_df)
cosine_distance

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

In [15]:
cosine_distance = pd.DataFrame(cosine_distance)

In [16]:
cosine_distance

0      1      2      3      4      5      6      7      8      9      \
0        1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1        0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2        0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
3        0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0   
4        0.0    0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
10828    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
10829    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
10830    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
10831    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
10832    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

       ...  10823  10824  10825  10826  10827  10828  10829  10830  10831  \
0      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
3      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
4      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
10828  ...    0.0    0.0    0.0    0.0    0.0    1.0    0.0    0.0    0.0   
10829  ...    0.0    0.0    0.0    0.0    0.0    0.0    1.0    0.0    0.0   
10830  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    1.0    0.0   
10831  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    1.0   
10832  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

       10832  
0        0.0  
1        0.0  
2        0.0  
3        0.0  
4        0.0  
...      ...  
10828    0.0  
10829    0.0  
10830    0.0  
10831    0.0  
10832    1.0  

[10833 rows x 10833 columns]

In [17]:
first_row = cosine_distance.iloc[0]
# Find indices and values of non-zero columns
non_zero_columns = first_row[first_row != 0]
non_zero_columns

0        1.000000
617      0.026955
2819     0.237826
4695     0.237826
4963     0.336336
7823     0.336336
8261     0.336336
8549     0.237826
8850     0.470871
10271    0.088326
Name: 0, dtype: float64

In [18]:
sorted_df = non_zero_columns.sort_values(ascending=False)
sorted_df

0        1.000000
8850     0.470871
4963     0.336336
7823     0.336336
8261     0.336336
2819     0.237826
8549     0.237826
4695     0.237826
10271    0.088326
617      0.026955
Name: 0, dtype: float64

In [19]:
first_row = cosine_distance.iloc[0]
# Find indices and values of non-zero columns
non_zero_columns = first_row[first_row != 1]
sorted_df = non_zero_columns.sort_values(ascending=False)
sorted_df

0        1.000000
8850     0.470871
7823     0.336336
8261     0.336336
4963     0.336336
           ...   
3613     0.000000
3614     0.000000
3615     0.000000
3616     0.000000
10832    0.000000
Name: 0, Length: 10833, dtype: float64

### **`We can see that User 8 and User 2168 shows a high corelation so lets check the books they are reading and what recomendation should given to both of them looking at their common interest`**

In [20]:
normalized_df.iloc[2168][normalized_df.iloc[2168]>0]

Book-Title
The Phantom Tollbooth    9.0
Name: 50650, dtype: float64

In [21]:
normalized_df.iloc[8][normalized_df.iloc[8]>0]

Book-Title
Seabiscuit: An American Legend    8.052632
To Kill a Mockingbird             8.947368
Name: 26, dtype: float64

In [37]:
first_half.index[first_half['User-ID'] == 8][0]

0

In [38]:
def userBasedCollaborativeFilter(user_id):
  # indexes to navigate in user-book rating matrix
  # indexes = normalized_df.index
  index = first_half.index[first_half['User-ID'] == user_id][0]

  # fetching the books that are already read be the user
  books_read_by_user = normalized_df.loc[user_id][normalized_df.loc[user_id]>0]

  # calculating the cosine distance of this user with all other users, and selecting user with highest cosine relation
  cos_dist = cosine_distance.iloc[index]
  non_zero_columns = cos_dist[cos_dist != 0]
  sorted_df = non_zero_columns.sort_values(ascending=False)
  users_with_similar_interest = sorted_df[sorted_df>.3].index
  books = list()
  for user in users_with_similar_interest[1:]:
     books.append(normalized_df.iloc[user][normalized_df.iloc[user]>0])

  merged_df = pd.concat(books)
  max_ratings = merged_df.groupby(level=0).max()
  max_ratings = max_ratings.sort_values(ascending=False)
  for i in books_read_by_user.index:
    try:
      max_ratings.drop(i, inplace = True)
    except:
      pass

  return max_ratings

In [44]:
recommend = userBasedCollaborativeFilter(200273)
recommend

Book-Title
Bant/Spec.Last of the Breed                                  9.000000
Piercing the Darkness                                        9.000000
The Rescue                                                   8.928571
Summer Island                                                8.052632
The Pilot's Wife : A Novel                                   8.052632
Skipping Christmas                                           8.000000
Midnight in the Garden of Good and Evil: A Savannah Story    7.941176
Prophet                                                      7.920000
Snow Falling on Cedars                                       7.142857
Death in the Clouds                                          7.058824
The Door to December                                         7.000000
From the Corner of His Eye                                   6.933333
The Clan of the Cave Bear (Earth's Children (Paperback))     6.066667
The Nanny Diaries: A Novel                                   5.280000
Slow Walt

In [22]:
normalized_df.to_csv('/content/drive/MyDrive/Colab Notebooks/Book Recommendation System/normalized_df.csv')
cosine_distance.to_csv('/content/drive/MyDrive/Colab Notebooks/Book Recommendation System/cosine_distance.csv')

In [27]:
def get_user(user_id = ''):
  if user_id:
    books_read_by_user = normalized_df.loc[user_id][normalized_df.loc[user_id]>0]
  else:
    books_read_by_user = normalized_df.head()
    books_read_by_user = first_half.sort_values(by='Book-Rating',ascending=False)
  return books_read_by_user.head(10)

In [28]:
get_user(8)

Book-Title
Clara Callan                                               4.102564
Goodbye to the Buttermilk Sky                              5.743590
Jane Doe                                                   4.102564
More Cunning Than Man: A Social History of Rats and Man    4.923077
The Middle Stories                                         4.102564
The Witchfinder (Amos Walker Mystery Series)               4.923077
Where You'll Find Me: And Other Stories                    4.102564
Name: 8, dtype: float64

In [ ]:
get_user(882)

Book-Title
A room with a view                                                                                                  8.840580
Across the Bitter Sea                                                                                               7.956522
Bridge to Terabithia                                                                                                8.840580
CADDIE WOODLAWN                                                                                                     8.840580
Dancing On My Grave Gelsey Kirkland                                                                                 7.956522
Diceys Song                                                                                                         7.956522
Eva Luna                                                                                                            8.840580
Farewell to Manzanar: A True Story of Japanese American Experience During and  After the World War II Internment  

In [ ]:
X = normalized_df.to_numpy()

In [ ]:
X

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [ ]:
from sklearn.decomposition import NMF

In [ ]:
model = NMF(n_components=2, init='random', random_state=0)

In [ ]:
W = model.fit_transform(X)

In [ ]:
H = model.components_

In [ ]:
W

array([[0.00325121, 0.00138479],
       [0.01196238, 0.00985759],
       [0.        , 0.        ],
       ...,
       [0.03135685, 0.016819  ],
       [0.03232558, 0.01002755],
       [0.01869817, 0.00810086]])

In [ ]:
H

array([[0.1701421 , 0.03891201, 0.00290061, ..., 0.00041601, 0.00570177,
        0.06876529],
       [0.145641  , 0.00174094, 0.02356109, ..., 0.0002287 , 0.03886875,
        0.03918858]])

In [ ]:
print('original matrix')

print(X)
print('---------------------------')
print('\n Factorized Matrix')

print(np.dot(W, H))

original matrix
[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]
---------------------------

 Factorized Matrix
[[7.54849183e-04 1.28921777e-04 4.20576747e-05 ... 1.66922508e-06
  7.23626921e-05 2.77838063e-04]
 [3.47097337e-03 4.82641800e-04 2.66953745e-04 ... 7.23086347e-06
  4.51358718e-04 1.20890147e-03]
 [0.00000000e+00 0.00000000e+00 0.00000000e+00 ... 0.00000000e+00
  0.00000000e+00 0.00000000e+00]
 ...
 [7.78465648e-03 1.24943907e-03 4.87228148e-04 ... 1.68911696e-05
  8.32522898e-04 2.81537582e-03]
 [6.96036479e-03 1.27531080e-03 3.30024117e-04 ... 1.57409609e-05
  5.74071266e-04 2.61584357e-03]
 [4.36116445e-03 7.41686689e-04 2.45101402e-04 ... 9.63123485e-06
  4.21483057e-04 1.60324676e-03]]


In [ ]:
factorized_array = np.dot(W,H)
factorized_array

array([[7.54849183e-04, 1.28921777e-04, 4.20576747e-05, ...,
        1.66922508e-06, 7.23626921e-05, 2.77838063e-04],
       [3.47097337e-03, 4.82641800e-04, 2.66953745e-04, ...,
        7.23086347e-06, 4.51358718e-04, 1.20890147e-03],
       [0.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00],
       ...,
       [7.78465648e-03, 1.24943907e-03, 4.87228148e-04, ...,
        1.68911696e-05, 8.32522898e-04, 2.81537582e-03],
       [6.96036479e-03, 1.27531080e-03, 3.30024117e-04, ...,
        1.57409609e-05, 5.74071266e-04, 2.61584357e-03],
       [4.36116445e-03, 7.41686689e-04, 2.45101402e-04, ...,
        9.63123485e-06, 4.21483057e-04, 1.60324676e-03]])

In [ ]:
# Create a copy of normalized dataframe and then paste the ratings from factorized matrix data into the copied dataframe
factorized_df = normalized_df.copy()
factorized_df

Book-Title  1984  2061: Odyssey Three  \
User-ID                                 
8            0.0                  0.0   
9            0.0                  0.0   
12           0.0                  0.0   
14           0.0                  0.0   
16           0.0                  0.0   
...          ...                  ...   
278681       0.0                  0.0   
278798       0.0                  0.0   
278832       0.0                  0.0   
278843       0.0                  0.0   
278851       0.0                  0.0   

Book-Title  28 Barbary Lane : A \Tales of the City\" Omnibus"  2nd Chance  \
User-ID                                                                     
8                                                         0.0         0.0   
9                                                         0.0         0.0   
12                                                        0.0         0.0   
14                                                        0.0         0.0   
16                                                        0.0         0.0   
...                                                       ...         ...   
278681                                                    0.0         0.0   
278798                                                    0.0         0.0   
278832                                                    0.0         0.0   
278843                                                    0.0         0.0   
278851                                                    0.0         0.0   

Book-Title  A Beautiful Mind : A Biography of John Forbes Nash, Jr.  \
User-ID                                                               
8                                                         0.0         
9                                                         0.0         
12                                                        0.0         
14                                                        0.0         
16                                                        0.0         
...                                                       ...         
278681                                                    0.0         
278798                                                    0.0         
278832                                                    0.0         
278843                                                    0.0         
278851                                                    0.0         

Book-Title  A Child Called \It\": One Child's Courage to Survive"  \
User-ID                                                             
8                                                         0.0       
9                                                         0.0       
12                                                        0.0       
14                                                        0.0       
16                                                        0.0       
...                                                       ...       
278681                                                    0.0       
278798                                                    0.0       
278832                                                    0.0       
278843                                                    0.0       
278851                                                    0.0       

Book-Title  A Christmas Story  A Cuisine of Leftovers  \
User-ID                                                 
8                         0.0                     0.0   
9                         0.0                     0.0   
12                        0.0                     0.0   
14                        0.0                     0.0   
16                        0.0                     0.0   
...                       ...                     ...   
278681                    0.0                     0.0   
278798                    0.0                     0.0   
278832                    0.0                     0.0   
278843         

In [ ]:
factorized_df.iloc[:] = factorized_array
factorized_df.head(10)

Book-Title           1984  2061: Odyssey Three  \
User-ID                                          
8            7.548492e-04         1.289218e-04   
9            3.470973e-03         4.826418e-04   
12           0.000000e+00         0.000000e+00   
14           2.948904e-06         1.892515e-07   
16           1.588314e-03         1.898613e-05   
17           2.274541e-03         4.026315e-04   
19           1.049689e-03         1.835909e-04   
22          1.049045e-237        1.253991e-239   
26           3.462385e-02         5.131899e-03   
32           6.881877e-04         8.226348e-06   

Book-Title  28 Barbary Lane : A \Tales of the City\" Omnibus"     2nd Chance  \
User-ID                                                                        
8                                                4.205767e-05   4.072945e-04   
9                                                2.669537e-04   2.110517e-03   
12                                               0.000000e+00   0.000000e+00   
14                                               3.742303e-07   2.269417e-06   
16                                               2.569496e-04   1.401283e-03   
17                                               1.172754e-04   1.196729e-03   
19                                               5.560521e-05   5.570762e-04   
22                                              1.697094e-238  9.255154e-238   
26                                               2.450983e-03   2.036814e-02   
32                                               1.113317e-04   6.071507e-04   

Book-Title  A Beautiful Mind : A Biography of John Forbes Nash, Jr.  \
User-ID                                                               
8                                                1.305761e-05         
9                                                9.123549e-05         
12                                               0.000000e+00         
14                                               1.400286e-07         
16                                               9.890641e-05         
17                                               3.533666e-05         
19                                               1.693663e-05         
22                                              6.532543e-239         
26                                               8.202226e-04         
32                                               4.285437e-05         

Book-Title  A Child Called \It\": One Child's Courage to Survive"  \
User-ID                                                             
8                                                7.566456e-04       
9                                                3.573463e-03       
12                                               0.000000e+00       
14                                               3.224827e-06       
16                                               1.807874e-03       
17                                               2.267843e-03       
19                                               1.048498e-03       
22                                              1.194059e-237       
26                                               3.537472e-02       
32                                               7.833191e-04       

Book-Title  A Christmas Story  A Cuisine of Leftovers  \
User-ID                                                 
8                4.830448e-08            3.364104e-08   
9                2.761707e-07            1.264435e-07   
12               0.000000e+00            0.000000e+00   
14               3.429659e-10            5.081688e-11   
16               2.254233e-07            6.104421e-09   
17               1.386056e-07            1.049987e-07   
19               6.505567e-08            4.788683e-08   
22              1.488870e-241           4.031831e-243   
26               2.599129e-06            1.342689e-06   
32               9.767188e-08            2.644936e-09   

Book-Title  A Dangerous Place: California's Unsettling Fate (Images of 

In [ ]:
get_user(56)

Book-Title
Haveli (Laurel Leaf Books)    7.875
LONESOME DOVE                 6.125
Name: 56, dtype: float64

In [ ]:
factorized_df.iloc[56].sort_values(ascending=False)

Book-Title
Harry Potter and the Chamber of Secrets (Book 2)       0.096203
Harry Potter and the Prisoner of Azkaban (Book 3)      0.090874
The Lovely Bones: A Novel                              0.076737
Harry Potter and the Goblet of Fire (Book 4)           0.052605
Harry Potter and the Sorcerer's Stone (Book 1)         0.041833
                                                         ...   
Going down swinging                                    0.000000
Tollivers Reisen. Stadtgeschichten IV.                 0.000000
The Vulcan Academy Murders                             0.000000
The Adventures of Drew and Ellie: The Magical Dress    0.000000
A Temporary New Wife                                   0.000000
Name: 226, Length: 817, dtype: float64

In [ ]:
recommend = userBasedCollaborativeFilter(56)
recommend

Book-Title
The Scarlet Letter                                                                               9.000000
The Time Traveler's Wife                                                                         8.823529
Snow Falling on Cedars                                                                           7.058824
The Pilot's Wife : A Novel Tag: Author of the Weight of Water (Oprah's Book Club (Hardcover))    7.058824
Dragonfly in Amber                                                                               6.909091
A Painted House                                                                                  6.176471
So You Want to Be a Wizard: The First Book in the Young Wizards Series                           5.666667
The Street Lawyer                                                                                5.181818
Me Talk Pretty One Day                                                                           5.000000
Chosen Prey                        